### PR TITLE
Replace cURL with a PSR-18 HTTP Client

### PR DESCRIPTION
**Pull request by JaZo**
I've replaced cURL with a PSR-18 HTTP Client, as the title says. Using raw cURL feels a little ancient and makes testing really hard. I also noted that SSL verification is currently turned off by default, which is a big no-no. By replacing cURL with a proper HTTP Client we can let the consumer decide all the settings (retry, SSL verification, timeout, etc.) without having to provide options for everything in this library. And it allows you to easily swap it for a mock client that doesn't make any HTTP calls in tests!

N.B. This PR isn't ready to be merged yet, but I saw in #286 that you were talking about a new major and this is a breaking change, so I wanted to bring it under your attention for consideration before the major is released.

### HTTP Client
I've deliberately made this client decoupled from any HTTP Client with the help of [PSR-18 HTTP Client](https://www.php-fig.org/psr/psr-18/). This allows the consumer to use and configure any HTTP Client they like. There are two ways we can deal with the HTTP Client dependency and we/you must decide which direction we should go.

1. Let this package require an HTTP Client, the most common is `guzzlehttp/guzzle`. This makes consuming this library the easiest. The drawback is that it installs Guzzle for users that provide their own HTTP Client and don't use Guzzle.
2. Let this package require `psr/http-client-implementation`. This is a virtual package telling Composer that a compatible client should be installed. We can then use `php-http/discovery` to automatically discover an available HTTP Client. This requires the consumer to separately install an HTTP Client e.g. `composer require mautic/api-library guzzlehttp/guzzle`. This circumvents the drawback of option 1, but makes installing this library a little more complicated.

### TODO
To summarize what needs to be done:
* [ ] Decide how we want to deal with the HTTP Client dependency.
* [ ] Update tests.

### CODE REVIEW

**Comment by JaZo** [comment]
I've signed the CLA yesterday, but it has probably not yet been processed.

**Comment by JaZo** [comment]
CC @Rocksheep

**Comment by RCheesley** [comment]
@cla-bot check 

Thanks for signing the PR @JaZo - just added you to the list.

Thanks also for the PR - welcome to the community!

**Comment by escopecz** [comment]
Thanks for this! It looks very good. And I like the second option. We can document the installation steps in the readme.

**Comment by JaZo** [comment]
@escopecz, sorry it took a while, but I've finally found the time to finish this up! It should be ready for review now. Only problem I had was running the tests locally. I tried running the tests on the main branch first, but that gives me 9 failing tests and I don't know Mautic enough to find out why. They may run just fine on CI...?

Steps I took:
1. Clone project (main branch);
2. `ddev start` and let it create the Mautic instance;
3. `ddev composer test`.

<details>
  <summary>Test output</summary>

```
There were 9 failures:

1) Mautic\Tests\Api\ContactFieldsTest::testGetList
The response has unexpected status code (500).

Response: {"errors":[{"message":"ErrorException: PHP Warning - mkdir(): File exists","code":500,"type":null}],"trace":[{"namespace":"","short_class":"","class":"","type":"","function":"","file":"\/var\/www\/html\/mautic\/vendor\/symfony\/http-kernel\/Profiler\/FileProfilerStorage.php","line":42},{"namespace":"Mautic\\CoreBundle\\ErrorHandler","short_class":"ErrorHandler","class":"Mautic\\CoreBundle\\ErrorHandler\\ErrorHandler","type":"->","function":"handleError","file":null,"line":null},{"namespace":"","short_class":"","class":"","type":"","function":"mkdir","file":"\/var\/www\/html\/mautic\/vendor\/symfony\/http-kernel\/Profiler\/FileProfilerStorage.php","line":42},{"namespace":"Symfony\\Component\\HttpKernel\\Profiler","short_class":"FileProfilerStorage","class":"Symfony\\Component\\HttpKernel\\Profiler\\FileProfilerStorage","type":"->","function":"__construct","file":"\/var\/www\/html\/mautic\/var\/cache\/dev\/ContainerPf8bcNh\/appAppKernelDevDebugContainer.php","line":5630},{"namespace":"Co...

Failed asserting that false is true.

/var/www/html/tests/Api/MauticApiTestCase.php:89
/var/www/html/tests/Api/MauticApiTestCase.php:146
/var/www/html/tests/Api/ContactFieldsTest.php:28
/var/www/html/tests/Api/MauticApiTestCase.php:203
/var/www/html/tests/Api/ContactFieldsTest.php:43

2) Mautic\Tests\Api\CategoriesTest::testGetList
The response has unexpected status code (500).

Response: {"errors":[{"message":"ErrorException: PHP Warning - mkdir(): File exists","code":500,"type":null}],"trace":[{"namespace":"","short_class":"","class":"","type":"","function":"","file":"\/var\/www\/html\/mautic\/vendor\/symfony\/cache\/Traits\/FilesystemCommonTrait.php","line":42},{"namespace":"Mautic\\CoreBundle\\ErrorHandler","short_class":"ErrorHandler","class":"Mautic\\CoreBundle\\ErrorHandler\\ErrorHandler","type":"->","function":"handleError","file":null,"line":null},{"namespace":"","short_class":"","class":"","type":"","function":"mkdir","file":"\/var\/www\/html\/mautic\/vendor\/symfony\/cache\/Traits\/FilesystemCommonTrait.php","line":42},{"namespace":"Symfony\\Component\\Cache\\Adapter","short_class":"PhpFilesAdapter","class":"Symfony\\Component\\Cache\\Adapter\\PhpFilesAdapter","type":"->","function":"init","file":"\/var\/www\/html\/mautic\/vendor\/symfony\/cache\/Adapter\/PhpFilesAdapter.php","line":33},{"namespace":"Symfony\\Component\\Cache\\Adapter","short_class":"PhpFiles...

Failed asserting that false is true.

/var/www/html/tests/Api/MauticApiTestCase.php:89
/var/www/html/tests/Api/MauticApiTestCase.php:146
/var/www/html/tests/Api/MauticApiTestCase.php:203
/var/www/html/tests/Api/CategoriesTest.php:26

3) Mautic\Tests\Api\SegmentsTest::testGetListOfSpecificIds
The response has unexpected status code (400).

Response: {"errors":[{"code":400,"message":"This form should not contain extra fields., filter: A value is required.","details":{"0":["This form should not contain extra fields."],"filter":["A value is required."]}}]}

Failed asserting that false is true.

/var/www/html/tests/Api/MauticApiTestCase.php:89
/var/www/html/tests/Api/SegmentsTest.php:55

4) Mautic\Tests\Api\SegmentsTest::testCreateGetAndDelete
The response has unexpected status code (400).

Response: {"errors":[{"code":400,"message":"This form should not contain extra fields., filter: A value is required.","details":{"0":["This form should not contain extra fields."],"filter":["A value is required."]}}]}

Failed asserting that false is true.

/var/www/html/tests/Api/MauticApiTestCase.php:89
/var/www/html/tests/Api/MauticApiTestCase.php:104
/var/www/html/tests/Api/SegmentsTest.php:84

5) Mautic\Tests\Api\SegmentsTest::testEditPatch
The response has unexpected status code (400).

Response: {"errors":[{"code":400,"message":"This form should not contain extra fields., filter: A value is required.","details":{"0":["This form should not contain extra fields."],"filter":["A value is required."]}}]}

Failed asserting that false is true.

/var/www/html/tests/Api/MauticApiTestCase.php:89
/var/www/html/tests/Api/MauticApiTestCase.php:104
/var/www/html/tests/Api/SegmentsTest.php:103

6) Mautic\Tests\Api\SegmentsTest::testEditPut
The response has unexpected status code (400).

Response: {"errors":[{"code":400,"message":"This form should not contain extra fields., filter: A value is required.","details":{"0":["This form should not contain extra fields."],"filter":["A value is required."]}}]}

Failed asserting that false is true.

/var/www/html/tests/Api/MauticApiTestCase.php:89
/var/www/html/tests/Api/MauticApiTestCase.php:104
/var/www/html/tests/Api/SegmentsTest.php:120

7) Mautic\Tests\Api\SegmentsTest::testAddAndRemove
The response has unexpected status code (400).

Response: {"errors":[{"code":400,"message":"This form should not contain extra fields., filter: A value is required.","details":{"0":["This form should not contain extra fields."],"filter":["A value is required."]}}]}

Failed asserting that false is true.

/var/www/html/tests/Api/MauticApiTestCase.php:89
/var/www/html/tests/Api/MauticApiTestCase.php:104
/var/www/html/tests/Api/SegmentsTest.php:137

8) Mautic\Tests\Api\SegmentsTest::testBatchEndpoints
This form should not contain extra fields., filter: A value is required.; This form should not contain extra fields., filter: A value is required.; This form should not contain extra fields., filter: A value is required.

Failed asserting that false is true.

/var/www/html/tests/Api/MauticApiTestCase.php:89
/var/www/html/tests/Api/MauticApiTestCase.php:104
/var/www/html/tests/Api/MauticApiTestCase.php:266
/var/www/html/tests/Api/SegmentsTest.php:173

9) Mautic\Tests\Api\WebhooksTest::testRemovingTriggerFromWebhook
The response has unexpected status code (500).

Response: {"errors":[{"message":"An exception occurred while executing \u0027UPDATE webhooks SET date_modified = ?, modified_by = ?, modified_by_user = ? WHERE id = ?\u0027 with params [\u00222022-12-27 20:12:09\u0022, 1, \u0022Admin Mautic\u0022, 12]:\n\nSQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction","code":500,"type":null}],"trace":[{"namespace":"","short_class":"","class":"","type":"","function":"","file":"\/var\/www\/html\/mautic\/vendor\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Driver\/AbstractMySQLDriver.php","line":51,"args":[]},{"namespace":"Doctrine\\DBAL\\Driver","short_class":"AbstractMySQLDriver","class":"Doctrine\\DBAL\\Driver\\AbstractMySQLDriver","type":"-\u003E","function":"convertException","file":"\/var\/www\/html\/mautic\/vendor\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/DBALException.php","line":182,"args":[["string","An exception occurred while executing \u0027UPDATE webhooks SET date_modified = ?, modified_by = ?, mod...

Failed asserting that false is true.

/var/www/html/tests/Api/MauticApiTestCase.php:89
/var/www/html/tests/Api/MauticApiTestCase.php:104
/var/www/html/tests/Api/WebhooksTest.php:107

FAILURES!
Tests: 304, Assertions: 4841, Failures: 9, Skipped: 7.
```
</details>

Env:
Ubuntu 22.04 on WSL2 (Windows 11)
DDEV 1.21.4

**Comment by escopecz** [comment]
I approved the test suite to run in the CI. Let's see

**Comment by JaZo** [comment]
FWIW: I tried running the tests locally again, but now it fails setting up Mautic. It looks like the [Mautic-version](https://github.com/mautic/mautic/tree/features) it tries to install is pretty outdated and doesn't support PHP 8.

<details>
<summary>ddev start output</summary>

```
Your lock file does not contain a compatible set of packages. Please run composer update.

  Problem 1
    - friendsofsymfony/oauth-server-bundle is locked to version dev-doctrine-fix and an update of this package was not requested.
    - friendsofsymfony/oauth-server-bundle dev-doctrine-fix requires php ^5.5|^7.0 -> your php version (8.0.28) does not satisfy that requirement.
  Problem 2
    - joomla/filter is locked to version 1.3.5 and an update of this package was not requested.
    - joomla/filter 1.3.5 requires php ^5.3.10|~7.0 -> your php version (8.0.28) does not satisfy that requirement.
  Problem 3
    - joomla/string is locked to version 1.4.5 and an update of this package was not requested.
    - joomla/string 1.4.5 requires php ^5.3.10|~7.0 -> your php version (8.0.28) does not satisfy that requirement.
  Problem 4
    - friendsofphp/php-cs-fixer is locked to version v2.16.10 and an update of this package was not requested.
    - friendsofphp/php-cs-fixer v2.16.10 requires php ^5.6 || ^7.0 -> your php version (8.0.28) does not satisfy that requirement.
  Problem 5
    - friendsofsymfony/oauth-server-bundle dev-doctrine-fix requires php ^5.5|^7.0 -> your php version (8.0.28) does not satisfy that requirement.
    - mautic/core-lib 3.2.x-dev requires friendsofsymfony/oauth-server-bundle dev-doctrine-fix -> satisfiable by friendsofsymfony/oauth-server-bundle[dev-doctrine-fix].
    - mautic/core-lib is locked to version 3.2.x-dev and an update of this package was not requested.
```
</details>

**Comment by JaZo** [comment]
I fixed the tests that failed in CI.

File: README.md [hunk 1]
```diff
@@ -16,6 +15,38 @@ You can install the API Library with the following command:
...
+use Mautic\Auth\ApiAuth;
+
+// Initiate an HTTP Client
+$httpClient = new Client([
+    'timeout'  => 10,
+]);
+
+// Initiate the auth object
+$initAuth = new ApiAuth();
+$auth     = $initAuth->newAuth($settings)->withHttpClient($httpClient);
```
Comment (escopecz) [inline README.md:46]: I'm thinking if this wouldn't be more dependency injection way:
```suggestion
$initAuth = new ApiAuth($httpClient);
$auth     = $initAuth->newAuth($settings);
```

File: composer.json [hunk 2]
```diff
@@ -15,17 +15,26 @@
...
+    "guzzlehttp/psr7": "^2.4",
+    "php-http/discovery": "^1.15"
   },
   "require-dev": {
     "kint-php/kint": "^4.1",
     "friendsofphp/php-cs-fixer": "~3.12.0",
     "phpunit/phpunit": "~9.5.0",
-    "brianium/paratest": "^6.2"
+    "brianium/paratest": "^6.2",
+    "guzzlehttp/guzzle": "^7.5"
```
Comment (escopecz) [inline composer.json:30]: I'm wondering whether this isn't a good candidate for Composer suggestion: https://getcomposer.org/doc/04-schema.md#suggest

File: README.md (hunk 1 above)
Comment (JaZo) [reply README.md]: Yes, that'd be a better solution, thanks!

File: lib/Auth/ApiAuth.php [hunk 3]
```diff
@@ -11,11 +11,23 @@
...
  * OAuth Client modified from https://code.google.com/p/simple-php-oauth/.
  */
 class ApiAuth
 {
+    /**
+     * @var \Psr\Http\Client\ClientInterface|null
+     */
+    protected $client;
+
+    public function __construct(ClientInterface $client = null)
```
Comment (escopecz) [inline lib/Auth/ApiAuth.php:26]: Is there any way this library will work without the Client object? I'm wondering whether making it nullable makes sense.

File: lib/Auth/ApiAuth.php [hunk 4]
```diff
@@ -11,11 +11,23 @@
...
+
 /**
  * OAuth Client modified from https://code.google.com/p/simple-php-oauth/.
  */
 class ApiAuth
 {
+    /**
+     * @var \Psr\Http\Client\ClientInterface|null
+     */
+    protected $client;
```
Comment (escopecz) [inline lib/Auth/ApiAuth.php:24]: Since we've dropped the PHP 7.4 support then we can start using typed properties

```php
protected ClientInterface $client
```

**Review by escopecz — CHANGES_REQUESTED** [review]
Thanks!

I recommend to merge https://github.com/mautic/api-library/pull/299 first as it will lower the changed rows significantly here.

**Comment by escopecz** [comment]
Also, please drop a note about this in https://github.com/mautic/api-library/blob/main/UPGRADE-4.0.md. Probably with a link to this PR for more details.

File: lib/Auth/ApiAuth.php (hunk 3 above)
Comment (JaZo) [reply lib/Auth/ApiAuth.php:26]: Yes, it will use auto-discovery to find an HTTP client when you pass `null`.

File: lib/Auth/ApiAuth.php (hunk 4 above)
Comment (JaZo) [reply lib/Auth/ApiAuth.php]: I've added types to all properties I've touched.

File: lib/Auth/ApiAuth.php (hunk 3 above)
Comment (escopecz) [reply lib/Auth/ApiAuth.php:23]: Interesting. It feels like magic. I don't like magic in code personally. But the property must be non-nullable, right?

Comment (JaZo) [reply lib/Auth/ApiAuth.php:23]: If we make it non-nullable, the user would be required to provide an HTTP client and they can't use auto-discovery. Or they would have to set up auto-discovery themselves. I've set up auto-discovery in the auth classes (in AbstractAuth to be precise) as these are the classes that actually need the HTTP client. However, these classes are usually instantiated with this factory class so I also had to add this argument and pass it through. I can move the auto-discovery set up to this class and make the HTTP client non-nullable in the auth classes if that feels less magic to you?

Comment (escopecz) [reply lib/Auth/ApiAuth.php:23]: What I want to avoid is to users of this library think that it'll work without any PSR HTTP Client. And that's what the nullable argument suggests.

Comment (mollux) [reply lib/Auth/ApiAuth.php:24]: The auto-discover in the class makes sense IMO.

@escopecz see also e.g. https://dev.to/rocksheep/how-to-use-psr-http-standards-to-upgrade-your-code-3l92#:~:text=)%3B%0A%20%20%20%20%7D%0A%7D-,Constructor,Discover%20classes%20to%20find%20usable%20classes%20like%20Guzzle%2C%20HTTPPlug%20or%20cURL.,-sendRequest

**Comment by mollux** [comment]
@JaZo I can't comment on an untracked file in a PR (for context: https://github.com/orgs/community/discussions/9099) so I add a comment here

IMO this need to be changed too:
https://github.com/mautic/api-library/blob/main/tests/Api/ContactsTest.php#L196-L201

I see 2 options:
* add `ext-curl` as dev dependency
* use the `psr/http-client` approach in the tests

**Comment by JaZo** [comment]
@mollux good catch! I could only find one usage of cURL in the tests, so I simply replaced it with a Guzzle call. Using auto-discovery there would not be possible as we have to configure the client to disable SSL verification.

**Comment by mollux** [comment]
@JaZo the tests are failing since the last change, can you look into it?

As you amended the commit, I can't easily see what the last change was. Can you please not amend commits to make the review process easier?

**Comment by JaZo** [comment]
@mollux, sorry for the amend, that's our way of working, so it's a force of habit. I've pushed the fixes for the tests in a separate commit!

**Comment by escopecz** [comment]
Sounds good. Thanks @mollux !

**Review by escopecz — APPROVED** [review]
